File: pkg/bench_nav/src/bench_nav/stepped/bug/visbug21.py
"""VisBug-21 — m-line sensor jumps, stepped."""

from __future__ import annotations

from collections.abc import Callable
from enum import Enum
from typing import override

from bench_nav.common import INF
from bench_nav.precomputation import COST
from bench_nav.stepped.bug._common import (
    WallFollowState,
    WallStepOutcome,
    bresenham,
    dir_to_goal,
    dist_sq,
    has_los,
    make_passable_closures,
    neighbour,
    sensed_cells,
    wall_follow_step,
)
from bench_nav.types import PrecompCtx, Stepped


class _Mode(Enum):
    MOTION = 0
    FOLLOW = 1


def _on_baseline(p: tuple[int, int], s: tuple[int, int], g: tuple[int, int]) -> bool:
    dx_t = g[0] - s[0]
    dy_t = g[1] - s[1]
    dx_c = p[0] - s[0]
    dy_c = p[1] - s[1]
    cross = abs(dy_c * dx_t - dx_c * dy_t)
    tol = max(abs(dx_t), abs(dy_t)) // 2
    if cross > tol:
        return False
    dot = dx_c * dx_t + dy_c * dy_t
    return dot > 0 and dist_sq(p, g) < dist_sq(s, g)


def _jump_along_mline(
    pos: tuple[int, int],
    start: tuple[int, int],
    goal: tuple[int, int],
    passable: Callable[[int, int], bool],
) -> tuple[int, int] | None:
    cur_d = dist_sq(pos, goal)
    best: tuple[tuple[int, int], int] | None = None
    for c in sensed_cells(pos):
        if c == pos or not _on_baseline(c, start, goal) or not passable(*c):
            continue
        d = dist_sq(c, goal)
        if d >= cur_d or not has_los(pos, c, passable):
            continue
        if best is None or d < best[1]:
            best = (c, d)
    return best[0] if best is not None else None


class VisBug21(Stepped):
    REQUIRES = frozenset({COST})

    @override
    def __init__(self, ctx: PrecompCtx) -> None:
        self.w = ctx.w
        self.h = ctx.h
        self.cost = ctx[COST]
        self.passable, self.on_map = make_passable_closures(
            ctx.w, ctx.h, self.cost, INF
        )
        self._active_goal: int | None = None
        self._start_xy: tuple[int, int] = (0, 0)
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(
            pos=(0, 0), current_obstacle=(0, 0), obstacle_on_right=True
        )
        self._hit_dist_sq = 0
        self._follow_visited: set[tuple[tuple[int, int], tuple[int, int], bool]] = set()
        self._los_queue: list[tuple[int, int]] = []

    def _reset(self, p: tuple[int, int], g: tuple[int, int]) -> None:
        self._start_xy = p
        self._mode = _Mode.MOTION
        self._wf = WallFollowState(pos=p, current_obstacle=p, obstacle_on_right=True)
        self._hit_dist_sq = dist_sq(p, g)
        self._follow_visited.clear()
        self._los_queue.clear()

    @override
    def step(self, pos: int, goal: int) -> int | None:
        w = self.w
        p = (pos % w, pos // w)
        g = (goal % w, goal // w)
        if goal != self._active_goal:
            self._active_goal = goal
            self._reset(p, g)
        if not self.passable(*p) or not self.passable(*g):
            return None
        if p == g:
            return pos
        if self._los_queue:
            nxt = self._los_queue.pop(0)
            return nxt[1] * w + nxt[0]

        if self._mode is _Mode.MOTION:
            target = _jump_along_mline(p, self._start_xy, g, self.passable)
            if target is not None:
                queue = bresenham(p, target)[1:]
                self._los_queue = queue[1:]
                first = queue[0]
                return first[1] * w + first[0]
            d = dir_to_goal(p, g)
            np = neighbour(p, d)
            if self.passable(*np):
                return np[1] * w + np[0]
            self._mode = _Mode.FOLLOW
            self._wf = WallFollowState(
                pos=p,
                current_obstacle=neighbour(p, d),
                obstacle_on_right=True,
            )
            self._hit_dist_sq = dist_sq(p, g)
            self._follow_visited.clear()
            self._follow_visited.add(
                (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
            )
            return pos
        # FOLLOW
        if dist_sq(p, g) <= 2 and self.passable(*g):
            self._mode = _Mode.MOTION
            return g[1] * w + g[0]
        self._wf.pos = p
        outcome = wall_follow_step(self._wf, self.passable, self.on_map)
        if outcome is WallStepOutcome.SURROUNDED:
            return None
        np = self._wf.pos
        if _on_baseline(np, self._start_xy, g) and dist_sq(np, g) < self._hit_dist_sq:
            self._mode = _Mode.MOTION
        state = (self._wf.pos, self._wf.current_obstacle, self._wf.obstacle_on_right)
        if state in self._follow_visited:
            return None
        self._follow_visited.add(state)
        return np[1] * w + np[0]

File: crates/titan-bugnav/src/algorithms/visbug21.rs
//! VisBug-21 (Lumelsky & Skewis, 1990). Bug2 with range-sensor jumps —
//! motion-to-goal jumps along the direction-to-goal ray; wall-follow takes
//! a single anchored step (no multi-cell wall jump, which overshoots
//! corners). The sensor shortcut in motion-to-goal gives Bug2's path the
//! main speedup.

use std::collections::HashSet;

use crate::algorithms::bug_common::{
    VISION_R_SQ, WallFollowState, WallStepOutcome, bresenham, dir_to_goal, dist_sq, has_los,
    neighbour, sensed_cells, wall_follow_step,
};
use crate::grid::Grid;
use crate::pathfinder::{Pathfinder, Snapshot, StepStatus};

fn on_baseline(pos: (i32, i32), start: (i32, i32), goal: (i32, i32)) -> bool {
    let dx_t = goal.0 - start.0;
    let dy_t = goal.1 - start.1;
    let dx_c = pos.0 - start.0;
    let dy_c = pos.1 - start.1;
    let cross = (dy_c * dx_t - dx_c * dy_t).abs();
    let tol = dx_t.abs().max(dy_t.abs()) / 2;
    if cross > tol {
        return false;
    }
    let dot = dx_c * dx_t + dy_c * dy_t;
    dot > 0 && dist_sq(pos, goal) < dist_sq(start, goal)
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum Mode {
    MotionToGoal,
    Follow,
}

pub struct VisBug21 {
    grid_w: i32,
    grid_h: i32,
    walls: Vec<bool>,
    start: (i32, i32),
    goal: (i32, i32),
    pos: (i32, i32),
    mode: Mode,
    wf: WallFollowState,
    hit_point: (i32, i32),
    hit_dist_sq: i32,
    follow_visited: HashSet<((i32, i32), (i32, i32), bool)>,
    steps: u32,
    snap: Snapshot,
    status: StepStatus,
}

#[must_use]
pub fn build(grid: &Grid, start: (i32, i32), goal: (i32, i32)) -> Box<dyn Pathfinder> {
    let mut snap = Snapshot {
        current: start,
        path: vec![start],
        ..Snapshot::default()
    };
    snap.visited.insert(start);
    Box::new(VisBug21 {
        grid_w: grid.w,
        grid_h: grid.h,
        walls: grid.walls.clone(),
        start,
        goal,
        pos: start,
        mode: Mode::MotionToGoal,
        wf: WallFollowState {
            pos: start,
            current_obstacle: start,
            obstacle_on_right: true,
        },
        hit_point: start,
        hit_dist_sq: dist_sq(start, goal),
        follow_visited: HashSet::new(),
        steps: 0,
        snap,
        status: StepStatus::Running,
    })
}

impl VisBug21 {
    fn passable(&self, x: i32, y: i32) -> bool {
        if x < 0 || y < 0 || x >= self.grid_w || y >= self.grid_h {
            return false;
        }
        !self.walls[(y * self.grid_w + x) as usize]
    }

    fn move_to(&mut self, new_pos: (i32, i32)) {
        self.pos = new_pos;
        self.snap.current = new_pos;
        self.snap.visited.insert(new_pos);
        self.snap.path.push(new_pos);
    }

    fn step_wall_follow(&mut self) -> WallStepOutcome {
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable =
            |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        let on_map = |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h;
        let outcome = wall_follow_step(&mut self.wf, passable, on_map);
        if outcome == WallStepOutcome::Moved {
            self.move_to(self.wf.pos);
        }
        outcome
    }

    fn enter_follow(&mut self, blocked_dir: usize) {
        self.mode = Mode::Follow;
        self.wf = WallFollowState {
            pos: self.pos,
            current_obstacle: neighbour(self.pos, blocked_dir),
            obstacle_on_right: true,
        };
        self.hit_point = self.pos;
        self.hit_dist_sq = dist_sq(self.pos, self.goal);
        self.follow_visited.clear();
        self.follow_visited.insert((
            self.wf.pos,
            self.wf.current_obstacle,
            self.wf.obstacle_on_right,
        ));
    }

    /// VisBug-21 motion-to-goal step: jump to the farthest visible cell on
    /// Bug2's planned path (the m-line from start to goal) that's closer to
    /// goal than current pos, with `LoS` and within sensor range. This is
    /// the paper's "farthest visible Ti on T" rule. Preserves m-line
    /// adherence, unlike a raw direction-to-goal jump which drifts off the
    /// baseline and breaks Follow's leave condition.
    fn jump_along_mline(&mut self) -> bool {
        let (w, h, walls) = (self.grid_w, self.grid_h, &self.walls);
        let passable =
            |x: i32, y: i32| x >= 0 && y >= 0 && x < w && y < h && !walls[(y * w + x) as usize];
        let cur_d = dist_sq(self.pos, self.goal);
        let mut best: Option<((i32, i32), i32)> = None;
        for c in sensed_cells(self.pos) {
            if c == self.pos {
                continue;
            }
            if !on_baseline(c, self.start, self.goal) {
                continue;
            }
            if !passable(c.0, c.1) {
                continue;
            }
            let d = dist_sq(c, self.goal);
            if d >= cur_d {
                continue;
            }
            if !has_los(self.pos, c, passable) {
                continue;
            }
            // Prefer the cell with the smallest dist-to-goal (farthest along
            // the m-line toward goal).
            match best {
                None => best = Some((c, d)),
                Some((_, bd)) if d < bd => best = Some((c, d)),
                _ => {}
            }
        }
        if let Some((target, _)) = best {
            let line = bresenham(self.pos, target);
            for &p in line.iter().skip(1) {
                self.move_to(p);
            }
            true
        } else {
            false
        }
    }
}

impl Pathfinder for VisBug21 {
    fn step(&mut self) -> StepStatus {
        if self.status != StepStatus::Running {
            return self.status;
        }
        if self.pos == self.goal {
            self.status = StepStatus::Arrived;
            return self.status;
        }
        self.steps += 1;

        match self.mode {
            Mode::MotionToGoal => {
                // Try the VisBug-21 m-line jump first.
                if self.jump_along_mline() {
                    return self.status;
                }
                // No visible m-line cell closer to goal — fall back to
                // Bug2's 1-cell direction-to-goal step.
                let d = dir_to_goal(self.pos, self.goal);
                let np = neighbour(self.pos, d);
                if self.passable(np.0, np.1) {
                    self.move_to(np);
                } else {
                    self.enter_follow(d);
                }
            }
            Mode::Follow => {
                if dist_sq(self.pos, self.goal) <= 2 && self.passable(self.goal.0, self.goal.1) {
                    self.mode = Mode::MotionToGoal;
                    self.move_to(self.goal);
                    return self.status;
                }
                match self.step_wall_follow() {
                    WallStepOutcome::Moved => {}
                    WallStepOutcome::Surrounded => {
                        self.status = StepStatus::Unreachable;
                        return self.status;
                    }
                }
                if on_baseline(self.pos, self.start, self.goal)
                    && dist_sq(self.pos, self.goal) < self.hit_dist_sq
                {
                    self.mode = Mode::MotionToGoal;
                    return self.status;
                }
                let state = (
                    self.wf.pos,
                    self.wf.current_obstacle,
                    self.wf.obstacle_on_right,
                );
                if !self.follow_visited.insert(state) {
                    self.status = StepStatus::Unreachable;
                }
            }
        }
        self.status
    }

    fn snapshot(&self) -> &Snapshot {
        &self.snap
    }

    fn summary(&self) -> String {
        format!(
            "pos: ({}, {})\ngoal: ({}, {})\nmode: {:?}\nhit: ({}, {})\nhit_d2: {}\nobstacle: ({}, {})\nside: {}\non_baseline: {}\nsensor r²≤{}\nsteps: {}\nstatus: {:?}",
            self.pos.0,
            self.pos.1,
            self.goal.0,
            self.goal.1,
            self.mode,
            self.hit_point.0,
            self.hit_point.1,
            self.hit_dist_sq,
            self.wf.current_obstacle.0,
            self.wf.current_obstacle.1,
            if self.wf.obstacle_on_right { "R" } else { "L" },
            on_baseline(self.pos, self.start, self.goal),
            VISION_R_SQ,
            self.steps,
            self.status,
        )
    }

    fn name(&self) -> &'static str {
        "VisBug-21"
    }
}
